pkg/gardenlet/controller/managedseed/managedseed_actuator.go

```diff
@@ -525,7 +548,52 @@ func (a *actuator) prepareGardenletChartValues(
 	gardenletConfig.SeedSelector = nil
 
 	// Get gardenlet chart values
-	return a.vp.GetGardenletChartValues(deployment, gardenletConfig, bootstrapKubeconfig)
+	return a.vp.GetGardenletChartValues(
+		ensureGardenletEnvironment(
+			deployment,
+			shoot.Spec.DNS,
+		),
+		gardenletConfig,
+		bootstrapKubeconfig,
+	)
+}
+
+// ensureGardenletEnvironment sets the KUBERNETES_SERVICE_HOST to the API of the ManagedSeed cluster.
+// This is needed so that the deployed gardenlet can properly set the network policies
+// allowing access of control plane components of the hosted shoots to the API server of the (managed) seed.
+func ensureGardenletEnvironment(deployment *seedmanagementv1alpha1.GardenletDeployment, dns *gardencorev1beta1.DNS) *seedmanagementv1alpha1.GardenletDeployment {
+	const (
+		kubernetesServiceHost = "KUBERNETES_SERVICE_HOST"
+	)
+	var (
+		shootDomain = ""
+	)
+
+	if deployment.Env == nil {
+		deployment.Env = []corev1.EnvVar{}
+	}
+
+	for i := len(deployment.Env) - 1; i >= 0; i-- {
+		if deployment.Env[i].Name == kubernetesServiceHost {
+			deployment.Env = append(deployment.Env[:i], deployment.Env[i+1:]...)
+		}
+	}
+
+	if dns != nil && dns.Domain != nil && len(*dns.Domain) != 0 {
+		shootDomain = fmt.Sprintf("api.%s", *dns.Domain)
+	}
+
+	if len(shootDomain) != 0 {
+		deployment.Env = append(
+			deployment.Env,
+			corev1.EnvVar{
+				Name:  kubernetesServiceHost,
+				Value: shootDomain,
+			},
+		)
+	}
```

Comment:
I think its better to change the implementation to one where you don't overwrite the `KUBERNETES_SERVICE_HOST` if it already exists. End-users might want to set it in the first place.

```suggestion
	if dns != nil && dns.Domain != nil && len(*dns.Domain) != 0 {
		shootDomain = fmt.Sprintf("api.%s", *dns.Domain)
	}

	if len(shootDomain) == 0 {
		return deployment
	}

	if deployment.Env == nil {
		deployment.Env = []corev1.EnvVar{}
	}

	for _, env := range deployment.Env {
		if env.Name == kubernetesServiceHost {
			// value is already set.
			return deployment
		}
	}

	deployment.Env = append(
		deployment.Env,
		corev1.EnvVar{
			Name:  kubernetesServiceHost,
			Value: shootDomain,
		},
	)
```

Reply:
Done.